pageserver/src/tenant/timeline.rs

```diff
@@ -1210,7 +1211,11 @@ impl Timeline {
                             anyhow::bail!("could not rename file {local_layer_path:?}: {err:?}");
                         } else {
                             self.metrics.resident_physical_size_gauge.sub(local_size);
-                            self.layers.write().unwrap().remove_historic(local_layer);
+                            {
+                                let mut layers = self.layers.write().unwrap();
+                                layers.remove_historic(local_layer);
+                                layers.rebuild_index();
```

Comment:
This rebuild (and the two rebuilds below) happen on every insertion. With 100k layers that's 0-100ms on every iteration. I'd prefer to rebuild after the for loop, but not sure what's the best way to do that. If I do it without holding a layer map lock, I risk someone seeing a deleted file in the layer map and panicking. But it might not be wise to hold a layer map lock if this method takes a while.

I'm not familiar with this code, would appreciate some help @SomeoneToIgnore

Reply:
On second thought, holding a write lock on the map while rebuilding is a more general problem I should address somehow before merging. Holding the lock for 100ms (on large dbs) doesn't seem fun

Reply:
Sorry, noticed the comment now only.
The main idea behind the code here is that on every timeline init we check the layers in the remote storage and trust that as a source of truth: all absent files are added as remote historic layers and all files that have a remote copy with different properties (now, simple size properly check only), gets replaced with the remote layer, local version gets removed.

`create_remote_layers` is supposed to happen (and I see it still does) only when pageserver has the remote storage enabled and 
* loads an existing timeline from disk (happens on pageserver start and `load` HTTP call
* loads a remote timeline from the remote storage during timeline attach (doubt, that's your case)

Based on that, I see two odd things: 

1. Why do we have "some" layer map that we reload? We can fill it properly from the first try, if we have a remote storage. If we can not do that (could be a big change, at least worth a separate PR), at least we could collect the changes and apply them once: could that help a bit?

2. Why is this a hot path for the tests? Odd, that something that happens once per timeline existence in pageserver's memory, has such a big impact. Might be not noticing a bigger picture.

Reply:
Thanks:
1. I didn't realize this is an initialization-only method. Should make things easier. We don't always have "some" pre-existing layer map, it's just that `rebuild_index()` is the `flush` method for batched updates. Maybe I'll rename it to `flush` for clarity
2. I don't know that it's a hot path. I didn't actually see any issues. The 100ms number comes from microbenchmarks, and I assumes it would cause issues. But if this is an init method It's not that big of a deal